File: 94.binary-tree-inorder-traversal.cpp
/*
 * @lc app=leetcode id=94 lang=cpp
 *
 * [94] Binary Tree Inorder Traversal
 */

#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    using NumsT = vector<int>;

    NumsT inorderTraversal(TreeNode* root)
    {
        NumsT res;
        Impl(root, res);
        return res;
    }

private:
    void Impl(TreeNode *n, NumsT &vals)
    {
        if (!n) return;
        Impl(n->left, vals);
        vals.push_back(n->val);
        Impl(n->right, vals);
    }
};
// @lc code=end

File: 94.binary-tree-inorder-traversal.py
#
# @lc app=leetcode id=94 lang=python3
#
# [94] Binary Tree Inorder Traversal
#

from typing import List, Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        def travers(n: TreeNode):
            if not n:
                return
            travers(n.left)
            res.append(n.val)
            travers(n.right)
        travers(root)
        return res
        
# @lc code=end

t1 = TreeNode(1, right=TreeNode(2, left=TreeNode(3)))
print(Solution().inorderTraversal(t1))